src/synchronous_product/include/synchronous_product/synchronous_product.h

```diff
@@ -0,0 +1,209 @@
+/** A TARegionState is a tuple (location, clock_name, clock_region) */
+template <typename Location>
+using TARegionState = std::tuple<Location, std::string, std::size_t>;
+/** An ATARegionState is a pair (formula, clock_region) */
+template <typename ActionType>
+using ATARegionState = std::pair<logic::MTLFormula<ActionType>, std::size_t>;
+/** An ABRegionSymbol is either a TARegionState or an ATARegionState */
+template <typename Location, typename ActionType>
+using ABRegionSymbol = std::variant<TARegionState<Location>, ATARegionState<ActionType>>;
+
+/** Get the clock valuation for an ABSymbol, which is either a TA state or an ATA state.
+ * @param w The symbol to read the time from
+ * @return The clock valuation in the given state
+ */
+template <typename Location, typename ActionType>
+ClockValuation
+get_time(const ABSymbol<Location, ActionType> &w)
+{
+	if (std::holds_alternative<TAState<Location>>(w)) {
+		return std::get<2>(std::get<TAState<Location>>(w));
+	} else {
+		return std::get<ATAState<ActionType>>(w).second;
+	}
+}
+
+/** Get the canonical word H(s) for the given A/B configuration s, closely
+ * following Bouyer et al., 2006. The TAStates of s are first expanded into
+ * triples (location, clock, valuation) (one for each clock), and then merged
+ * with the pairs from the ATAConfiguration. The resulting set is then
+ * partitioned according to the fractional part of the clock valuations. Then,
+ * each tuple is regionalized by replacing the clock valuation with the
+ * respective region index.  The resulting word is a sequence of sets, each set
+ * containing regionalized tuples that describe a TAState or ATAState. The
+ * sequence is sorted by the fractional part of the original clock valuations.
+ * @param ta_configuration The configuration of the timed automaton A
+ * @param ata_configuration The configuration of the alternating timed automaton B
+ * @param K The value of the largest constant any clock may be compared to
+ * @return The canonical word representing the state s, as a sorted vector of
+ * sets of tuples (triples from A and pairs from B).
+ */
+template <typename Location, typename ActionType>
+std::vector<std::set<ABRegionSymbol<Location, ActionType>>>
+get_canonical_word(const automata::ta::Configuration<Location> &ta_configuration,
+                   const ATAConfiguration<ActionType> &         ata_configuration,
+                   const unsigned int                           K)
+{
+	std::set<ABSymbol<Location, ActionType>> g;
+	// Insert ATA configurations into g.
+	std::copy(ata_configuration.begin(), ata_configuration.end(), std::inserter(g, g.end()));
+	// Insert TA configurations into g.
+	for (const auto &[clock_name, clock_value] : ta_configuration.second) {
+		g.insert(std::make_tuple(ta_configuration.first, clock_name, clock_value));
+	}
+	// Sort into partitions by the fractional parts.
+	std::map<double, std::set<ABSymbol<Location, ActionType>>> partitioned_g;
```

Comment:
I think double should be ClockValuation, right?

Reply:
Yes, thanks for the fix!